File: 004. Median of Two Sorted Arrays.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        if(nums1.size()> nums2.size()) return findMedianSortedArrays(nums2, nums1);
        int m = nums1.size(), n = nums2.size();
        int low = 0, high = m, i, half = (m+n+1)/2;
        
        while(low <=high){
            int i = (low+high)/2;
            int j = half - i;
            
            if(i>0 && nums1[i-1]>nums2[j]){
                high = i-1;
            }
            else if( i<m && nums2[j-1]>nums1[i]){
                low = i+1;
            }
            else{
                int max_left, min_right;
                if(i==0) max_left = nums2[j-1];
                else if( j ==0) max_left = nums1[i-1];
                else max_left = max(nums2[j-1], nums1[i-1]);
                
                if((m+n)%2 == 1) return max_left;
                
                if(i == m) min_right = nums2[j];
                else if(j == n) min_right = nums1[i];
                else min_right = min(nums2[j],nums1[i]);
                
                return (min_right + max_left)/2.0;
            }
        }
    }
};

File: 004_median_of_two_sorted_arrays.py
# a wrong sol
#ref: sol3 of https://leetcode.cn/problems/median-of-two-sorted-arrays/solutions/8999/xiang-xi-tong-su-de-si-lu-fen-xi-duo-jie-fa-by-w-2/
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        m, n = len(nums1), len(nums2)
        k = (m+n+1)//2
        last_1, last_2 = 0, 0

        is_on_1 = True
        while k >= 1:
            step = k//2
            if last_1 == m:
                last_2 += k-1
                is_on_1 = False
                break
            if last_2 == n:
                last_1 += k-1
                is_on_1 = True
                break
            if k == 1:
                if nums1[last_1] >= nums2[last_2]:
                    is_on_1 = False
                    break
                else:
                    is_on_1 = True
                    break
            if nums1[last_1 + step -1] >= nums2[last_2 + step -1]:
                last_2 += step
                is_on_1 = True
            else:
                last_1 += step
                is_on_1 = False
            k -= step
        if (m+n)%2 == 1:
            return nums1[last_1] if is_on_1 else nums2[last_2]
        if (m+n)%2 == 0:
            if last_1 == m:
                return (nums2[last_2] + nums2[last_2+1])/2
            if last_2 == n:
                return (nums1[last_1] + nums1[last_1+1])/2
            else:
                if is_on_1:
                    if last_1 == m-1:
                        return (nums1[last_1] + nums2[last_2])/2
                    return (nums1[last_1] + min(nums1[last_1+1], nums2[last_2]))/2
                else:
                    if last_2 == n-1:
                        return (nums1[last_1] + nums2[last_2])/2
                    return (nums2[last_2] + min(nums2[last_2+1], nums1[last_1]))/2
